router/pkg/pubsub/datasource/subscription_datasource.go

```diff
@@ -0,0 +1,75 @@
+package datasource
+
+import (
+	"encoding/json"
+	"errors"
+
+	"github.com/cespare/xxhash/v2"
+	"github.com/wundergraph/graphql-go-tools/v2/pkg/engine/resolve"
+)
+
+type uniqueRequestIdFn func(ctx *resolve.Context, input []byte, xxh *xxhash.Digest) error
+
+// PubSubSubscriptionDataSource is a data source for handling subscriptions using a Pub/Sub mechanism.
+// It implements the SubscriptionDataSource interface and HookableSubscriptionDataSource
+type PubSubSubscriptionDataSource[C SubscriptionEventConfiguration] struct {
+	pubSub                 Adapter
+	uniqueRequestID        uniqueRequestIdFn
+	subscriptionOnStartFns []SubscriptionOnStartFn
+}
+
+func (s *PubSubSubscriptionDataSource[C]) SubscriptionEventConfiguration(input []byte) (SubscriptionEventConfiguration, error) {
+	var subscriptionConfiguration C
+	err := json.Unmarshal(input, &subscriptionConfiguration)
+	if err != nil {
+		return nil, err
+	}
+	return subscriptionConfiguration, nil
```

Comment:
nit: I think you could either remove the if or inline the error

```suggestion
	var subscriptionConfiguration C
	err := json.Unmarshal(input, &subscriptionConfiguration)
	return subscriptionConfiguration, err
```